File: pc/src/host/BridgeEndpointPolicy.ts
export type BridgeEndpointSecurity = "secure" | "local-development" | "trusted-overlay-development";

export interface NormalizedBridgeEndpoint {
  url: string;
  security: BridgeEndpointSecurity;
}

export function normalizeBridgeEndpoint(
  raw: string,
  options: { allowInsecureTrustedOverlay?: boolean } = {}
): NormalizedBridgeEndpoint {
  let parsed: URL;
  try {
    parsed = new URL(raw.trim());
  } catch {
    throw new Error("Bridge endpoint must be a valid WebSocket URL.");
  }
  if (!new Set(["ws:", "wss:"]).has(parsed.protocol)
    || !parsed.hostname
    || parsed.username
    || parsed.password
    || parsed.search
    || parsed.hash
    || (parsed.pathname !== "/" && parsed.pathname !== "/phone")) {
    throw new Error("Bridge endpoint must be a credential-free ws or wss URL ending in /phone.");
  }
  parsed.pathname = "/phone";
  if (parsed.protocol === "wss:") {
    return { url: parsed.toString(), security: "secure" };
  }
  if (isLoopbackHost(parsed.hostname)) {
    return { url: parsed.toString(), security: "local-development" };
  }
  if (options.allowInsecureTrustedOverlay === true && isTrustedOverlayHost(parsed.hostname)) {
    return { url: parsed.toString(), security: "trusted-overlay-development" };
  }
  throw new Error("Network bridge endpoints require wss; ws is limited to loopback/ADB or an explicitly trusted Tailscale overlay.");
}

export function isLoopbackHost(host: string): boolean {
  const normalized = stripIpv6Brackets(host).toLowerCase();
  return normalized === "localhost" || normalized === "::1" || /^127(?:\.\d{1,3}){3}$/.test(normalized);
}

export function isTrustedOverlayHost(host: string): boolean {
  const normalized = stripIpv6Brackets(host).toLowerCase().replace(/\.$/, "");
  if (normalized.endsWith(".ts.net") || normalized.startsWith("fd7a:115c:a1e0:")) {
    return true;
  }
  const octets = normalized.split(".").map(Number);
  return octets.length === 4
    && octets.every((value) => Number.isInteger(value) && value >= 0 && value <= 255)
    && octets[0] === 100
    && octets[1]! >= 64
    && octets[1]! <= 127;
}

function stripIpv6Brackets(host: string): string {
  return host.startsWith("[") && host.endsWith("]") ? host.slice(1, -1) : host;
}

File: android/app/src/main/java/dev/androidagent/BridgeEndpointPolicy.kt
package dev.androidagent

import java.net.URI

enum class BridgeEndpointSecurity {
    Secure,
    LocalDevelopment,
    TrustedOverlayDevelopment
}

data class NormalizedBridgeEndpoint(val url: String, val security: BridgeEndpointSecurity)

object BridgeEndpointPolicy {
    fun normalize(raw: String, allowInsecureTrustedOverlay: Boolean = false): NormalizedBridgeEndpoint? {
        val uri = runCatching { URI(raw.trim()) }.getOrNull() ?: return null
        val scheme = uri.scheme?.lowercase() ?: return null
        val host = uri.host?.lowercase()?.trimEnd('.') ?: return null
        if (scheme !in setOf("ws", "wss") || uri.userInfo != null || uri.fragment != null || uri.query != null) return null
        if (uri.port !in -1..65_535) return null
        if (!uri.path.isNullOrEmpty() && uri.path != "/" && uri.path != "/phone") return null
        val normalized = URI(scheme, null, host, uri.port, "/phone", null, null).toString()
        return when {
            scheme == "wss" -> NormalizedBridgeEndpoint(normalized, BridgeEndpointSecurity.Secure)
            isLoopback(host) -> NormalizedBridgeEndpoint(normalized, BridgeEndpointSecurity.LocalDevelopment)
            allowInsecureTrustedOverlay && isTrustedOverlay(host) ->
                NormalizedBridgeEndpoint(normalized, BridgeEndpointSecurity.TrustedOverlayDevelopment)
            else -> null
        }
    }

    fun protectsProviderSecrets(raw: String): Boolean =
        normalize(raw)?.security == BridgeEndpointSecurity.Secure

    private fun isLoopback(host: String): Boolean =
        host == "localhost" || host == "::1" || Regex("^127(?:\\.\\d{1,3}){3}$").matches(host)

    private fun isTrustedOverlay(host: String): Boolean {
        if (host.endsWith(".ts.net") || host.startsWith("fd7a:115c:a1e0:")) return true
        val octets = host.split('.').mapNotNull(String::toIntOrNull)
        return octets.size == 4 && octets.all { it in 0..255 } && octets[0] == 100 && octets[1] in 64..127
    }
}
